### PR TITLE
Bufferparam

### PR DESCRIPTION
**Pull request by tokazio**
Changed buffering system to make number and size of the streaming buffers editable.

### CODE REVIEW

File: src/main/java/org/urish/openal/Buffer.java [hunk 1]
```diff
@@ -1,99 +1,193 @@
...
+	return "ALBuffer[" + getBufferId() + "]";
+    }
+
+    /**
+     *
+     * @param param
+     * @return
+     * @throws ALException
+     */
+    public int getIntParam(int param) throws ALException {
```
Comment (tokazio) [inline src/main/java/org/urish/openal/Buffer.java:187]: Like in Source but for buffer infos

File: src/main/java/org/urish/openal/Source.java [hunk 2]
```diff
@@ -12,291 +9,517 @@
...
+	setFloat3Param(AL.AL_VELOCITY, velocity);
+    }
+
+    /**
+     *
+     * @param format
+     * @return
+     * @throws ALException
+     */
+    public SourceBufferedOutputStream createOutputStream(AudioFormat format) throws ALException {
```
Comment (tokazio) [inline src/main/java/org/urish/openal/Source.java:469]: For compatibility

File: src/main/java/org/urish/openal/Source.java [hunk 3]
```diff
@@ -12,291 +9,517 @@
...
+    
+    /**
+     *
+     * @param format
+     * @param numberOfBuffer
+     * @param aBufferSize
+     * @return
+     * @throws ALException
+     */
+    public SourceBufferedOutputStream createOutputStream(AudioFormat format, int numberOfBuffer, int aBufferSize) throws ALException {
```
Comment (tokazio) [inline src/main/java/org/urish/openal/Source.java:481]: Number of buffer + Buffer size

Return SourceBufferedOutputStream to get access to samplesProcessed that can give ms ellapsed time

File: src/main/java/org/urish/openal/SourceBufferedOutputStream.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
+package org.urish.openal;
+
+import java.io.BufferedOutputStream;
+
+/**
+ *
+ * @author Romain PETIT <tokazio@esyo.net>
+ */
+public class SourceBufferedOutputStream extends BufferedOutputStream{
```
Comment (tokazio) [inline src/main/java/org/urish/openal/SourceBufferedOutputStream.java:9]: SourceOutputStream as Buffered

File: src/main/java/org/urish/openal/SourceOutputStream.java [hunk 5]
```diff
@@ -13,71 +13,188 @@
...
+public class SourceOutputStream extends OutputStream {
 
-	@Override
-	public void write(int b) throws IOException {
-		write(new byte[] { (byte) b }, 0, 1);
-	}
+    /**
+     * 
+     */
+    private int CYCLIC_BUFFER_COUNT = 3;
```
Comment (tokazio) [inline src/main/java/org/urish/openal/SourceOutputStream.java:24]: 3 buffers by default

File: src/main/java/org/urish/openal/SourceOutputStream.java [hunk 6]
```diff
@@ -13,71 +13,188 @@
...
+     * 
+     * for compatibility
+     * @param al
+     * @param source
+     * @param format
+     * @throws ALException 
+     * @deprecated Use SourceOutputStream with the numberOfBuffer param
+     */
+    @Deprecated
+    public SourceOutputStream(AL al, Source source, AudioFormat format) throws ALException {
```
Comment (tokazio) [inline src/main/java/org/urish/openal/SourceOutputStream.java:71]: For compatibility

File: src/main/java/org/urish/openal/SourceOutputStream.java [hunk 7]
```diff
@@ -13,71 +13,188 @@
...
+    
+    /**
+     * 
+     * @param al
+     * @param source
+     * @param format
+     * @param numberOfBuffer
+     * @throws ALException 
+     */
+    public SourceOutputStream(AL al, Source source, AudioFormat format, int numberOfBuffer) throws ALException {
```
Comment (tokazio) [inline src/main/java/org/urish/openal/SourceOutputStream.java:87]: With number of buffers

File: src/main/java/org/urish/openal/SourceOutputStream.java [hunk 8]
```diff
@@ -13,71 +13,188 @@
...
+	this.format = format;
+	setNumberOfBuffer(numberOfBuffer);
+	initBuffers();
+    }
+
+    /**
+     * Set number and buffer and initialize them
+     * @param numberOfBuffer 
+     */
+    private void setNumberOfBuffer(int numberOfBuffer) throws ALException{
```
Comment (tokazio) [inline src/main/java/org/urish/openal/SourceOutputStream.java:100]: setter

File: src/main/java/org/urish/openal/SourceOutputStream.java [hunk 9]
```diff
@@ -13,71 +13,188 @@
...
+	CYCLIC_BUFFER_COUNT = numberOfBuffer;
+	initBuffers();
+    }
+    
+    /**
+     * 
+     * @return The number of buffer
+     */
+    public int getNumberOfBuffer(){
```
Comment (tokazio) [inline src/main/java/org/urish/openal/SourceOutputStream.java:109]: getter

File: src/main/java/org/urish/openal/SourceOutputStream.java [hunk 10]
```diff
@@ -13,71 +13,188 @@
...
+	return CYCLIC_BUFFER_COUNT;
+    }
+    
+    /**
+     * Initialize buffers (empty buffers)
+     */
+    private void initBuffers() throws ALException{
```
Comment (tokazio) [inline src/main/java/org/urish/openal/SourceOutputStream.java:116]: initializing buffer array with number of buffer

File: src/main/java/org/urish/openal/SourceOutputStream.java [hunk 11]
```diff
@@ -13,71 +13,188 @@
...
+	for (int i = 0; i < buffers.length; i++) {
+	    buffers[i] = new Buffer(al);
 	}
+    }
+    
+    /**
+     * samplesProcessed/sampleRate/channels -> ms
+     * @return 
+     */
+    public long getSamplesProcessed(){
```
Comment (tokazio) [inline src/main/java/org/urish/openal/SourceOutputStream.java:127]: getter

File: src/main/java/org/urish/openal/SourceOutputStream.java [hunk 12]
```diff
@@ -13,71 +13,188 @@
...
+		while (source.getProcessedBufferCount() == 0) {
+		    try {
+			Thread.sleep(20);
+		    } catch (InterruptedException e) {
+			throw new IOException("IO Operation interrupted", e);
+		    }
 		}
+		source.unqueueBuffer(buffers[queueTail]);
+
+		samplesProcessed += buffers[queueTail].getIntParam(AL.AL_SIZE) / (buffers[queueTail].getIntParam(AL.AL_BITS) / 8);
```
Comment (tokazio) [inline src/main/java/org/urish/openal/SourceOutputStream.java:172]: Compute number of samples processed:
samplesProcessed/sampleRate/channels can give milliseconds processed

**Comment by urish** [comment]
Thanks!

It is really hard to follow the diffs though, as you seem the have reformatted the entire code and added some comments...

**Comment by tokazio** [comment]
Yes, i've put comments where important things has changed.

**Comment by urish** [comment]
Thanks!

**Comment by tokazio** [comment]
Thanks to you, nice work! 
and it was my first pull request ;)

**Comment by urish** [comment]
Congrats! Keep it going :)
